packages/core/src/components/fields/LayoutGridField.tsx

```diff
@@ -140,6 +138,592 @@ function isNumericIndex(str: string) {
+    // `uiReadonly` flag that it returns. This is done since the `SchemaField` will always defer to the `readonly`
+    // state in the uiSchema over anything in the props or schema. Because we are implementing the "readonly" state of
+    // the `Form` via the prop passed to `LayoutGridField` we need to make sure the uiSchema always has a true value
+    // when it is needed
+    const { fieldUiSchema, uiReadonly } = computeFieldUiSchema<T, S, F>(name, uiProps, uiSchema, isReadonly, readonly);
+
+    return (
+      <Field
+        data-testid={
+          optionsInfo?.hasDiscriminator
+            ? LAYOUT_GRID_FIELD_TEST_IDS.layoutMultiSchemaField
+            : LAYOUT_GRID_FIELD_TEST_IDS.field
+        }
+        {...otherProps}
+        name={name}
+        required={isRequired}
+        readonly={uiReadonly}
+        schema={schema}
+        uiSchema={fieldUiSchema}
+        errorSchema={get(errorSchema, name)}
+        fieldPathId={fieldIdSchema}
+        formData={get(formData, name)}
+        onChange={onChange}
+        onBlur={onBlur}
+        onFocus={onFocus}
+        options={optionsInfo?.options}
+        registry={registry}
+      />
+    );
+  }
+
+  if (UIComponent) {
+    return (
+      <UIComponent
+        data-testid={LAYOUT_GRID_FIELD_TEST_IDS.uiComponent}
+        {...otherProps}
+        name={name}
+        required={isRequired}
+        formData={formData}
+        readOnly={!!isReadonly || readonly}
+        errorSchema={errorSchema}
+        uiSchema={uiSchema}
+        schema={initialSchema}
+        fieldPathId={fieldPathId}
+        onBlur={onBlur}
+        onFocus={onFocus}
+        registry={registry}
+        {...uiProps}
+      />
+    );
+  }
+  return null;
+}
+
+/** Computes the uiSchema for the field with `name` from the `uiProps` and `uiSchema` provided. The field UI Schema
+ * will always contain a copy of the global options from the `uiSchema` (so they can be passed down) as well as
+ * copying them into the local ui options. When the `forceReadonly` flag is true, then the field UI Schema is
+ * updated to make "readonly" be true. When the `schemaReadonly` flag is true AND the field UI Schema does NOT have
+ * the flag already provided, then we also make "readonly" true. We always make sure to return the final value of the
+ * field UI Schema's "readonly" flag as `uiReadonly` along with the `fieldUiSchema` in the return value.
+ *
+ * @param field - The name of the field to pull the existing UI Schema for
+ * @param uiProps - Any props that should be put into the field's uiSchema
+ * @param [uiSchema] - The optional UI Schema from which to get the UI schema for the field
+ * @param [schemaReadonly] - Optional flag indicating whether the schema indicates the field is readonly
+ * @param [forceReadonly] - Optional flag indicating whether the Form itself is in readonly mode
+ */
+export function computeFieldUiSchema<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
```

Comment:
Consider moving all the helper functions above the components that are using them...